src/core/hle/kernel/thread.cpp

```diff
@@ -3,433 +3,972 @@
+    }
 
-void ThreadManager::DebugThreadQueue() {
-    Thread* thread = GetCurrentThread();
-    if (!thread) {
-        LOG_DEBUG(Kernel, "Current: NO CURRENT THREAD");
-    } else {
-        LOG_DEBUG(Kernel, "0x{:02X} {} (current)", thread->current_priority,
-                  GetCurrentThread()->GetObjectId());
+    thread->waiting_on = objects;
+    for (auto obj : objects) {
+        obj->AddWaitingThread(Kernel::SharedFrom(thread));
     }
+    thread->status = Thread::Status::WaitIPC;
+    Reschedule();
+    return thread->SyncOutput(RESULT_SUCCESS);
+}
 
-    for (auto& t : thread_list) {
-        u32 priority = ready_queue.contains(t.get());
-        if (priority != -1) {
-            LOG_DEBUG(Kernel, "0x{:02X} {}", priority, t->GetObjectId());
+ResultCode ThreadManager::WaitAll(std::vector<ObjectPtr> objects,
+                                  std::chrono::nanoseconds timeout) {
+    ASSERT(thread->status == Thread::Status::Running);
+
+    // Check if we can sync immediately
+    auto inactive_obj = std::find_if(objects.begin(), objects.end(),
+                                     [this](const auto& obj) { return obj->ShouldWait(thread); });
+    if (inactive_obj == objects.end()) {
+        for (auto obj : objects) {
+            obj->Acquire(thread);
         }
+        return RESULT_SUCCESS;
     }
-}
 
-/**
- * Finds a free location for the TLS section of a thread.
- * @param tls_slots The TLS page array of the thread's owner process.
- * Returns a tuple of (page, slot, alloc_needed) where:
- * page: The index of the first allocated TLS page that has free slots.
- * slot: The index of the first free slot in the indicated page.
- * alloc_needed: Whether there's a need to allocate a new TLS page (All pages are full).
- */
-static std::tuple<std::size_t, std::size_t, bool> GetFreeThreadLocalSlot(
-    const std::vector<std::bitset<8>>& tls_slots) {
-    // Iterate over all the allocated pages, and try to find one where not all slots are used.
-    for (std::size_t page = 0; page < tls_slots.size(); ++page) {
-        const auto& page_tls_slots = tls_slots[page];
-        if (!page_tls_slots.all()) {
-            // We found a page with at least one free slot, find which slot it is
-            for (std::size_t slot = 0; slot < page_tls_slots.size(); ++slot) {
-                if (!page_tls_slots.test(slot)) {
-                    return std::make_tuple(page, slot, false);
-                }
-            }
+    // Schedule the timeout
+    if (timeout.count() == 0) {
+        auto active_obj = std::find_if(objects.begin(), objects.end(), [this](const auto& obj) {
+            return !obj->ShouldWait(thread);
+        });
+        if (active_obj == objects.end()) {
+            return RESULT_TIMEOUT;
         }
+    } else {
+        root.ScheduleEvent(thread->wakeup_event.get(), timeout);
     }
 
-    return std::make_tuple(0, 0, true);
+    // Wait on the objects
+    thread->waiting_on = objects;
+    for (auto obj : objects) {
```

Comment:
```suggestion
    for (auto& obj : objects) {
```

otherwise this can cause atomic reference count increment/decrement churn.